api_analysis/lib/r3/analyze.dart

```diff
@@ -0,0 +1,491 @@
+  do {
+    changed = false;
+    for (final lib in package.libraries.values) {
+      changed = changed || package.propagateExports(lib);
+    }
+  } while (changed);
+  return package;
+}
+
+extension on PackageShape {
+  void addLibrary(ParsedLibraryResult library) {
+    final libraryUnit = library.units.where((u) => u.isLibrary).firstOrNull;
+    if (libraryUnit == null) {
+      fail('Could not find libraryUnit for $library');
+    }
+    // Find the URI for this library
+    final uri = libraryUnit.uri;
+
+    // Check if library shape has been added before, this can happen if a
+    // library has multiple parts.
+    if (libraries[uri] != null) {
+      return;
+    }
+
+    final shape = libraries[uri] = LibraryShape(uri: uri);
+
+    // Extract import / export statements
+    for (final d in library.units.expand((u) => u.unit.directives)) {
+      switch (d) {
+        case ExportDirective d:
+          shape.addExport(d);
+
+        case ImportDirective d:
+          shape.addImport(d);
+
+        case LibraryDirective():
+        case PartDirective() || PartOfDirective():
+        case AugmentationImportDirective() || LibraryAugmentationDirective():
+      }
+    }
+
+    // Find all top-level declarations
+    for (final d in library.units.expand((u) => u.unit.declarations)) {
+      switch (d) {
+        case FunctionDeclaration d:
+          if (d.isGetter || d.isSetter) {
+            if (shape.definedShapes.containsKey(d.name.lexeme)) {
+              continue; // we've seen the other getter or setter already
+              // TODO: Check that the other thing we saw was a getter/setter!
+            }
+            final accessors = library.units
+                .expand((u) => u.unit.declarations)
+                .whereType<FunctionDeclaration>()
+                .where((dOther) => d.name.lexeme == dOther.name.lexeme);
+            shape.define(VariableShape(
+              name: d.name.lexeme,
+              hasGetter: accessors.any((a) => a.isGetter),
+              hasSetter: accessors.any((a) => a.isSetter),
+            ));
+          } else {
+            shape.defineFunction(d);
+          }
+
+        case EnumDeclaration d:
+          shape.defineEnum(d);
+
+        case ClassDeclaration d:
+          shape.defineClass(d);
+
+        case MixinDeclaration d:
+          shape.defineMixin(d);
+
+        case ExtensionDeclaration d:
+          shape.defineExtension(d);
+
+        case ClassTypeAlias d:
+          shape.defineClassTypeAlias(d);
+
+        case FunctionTypeAlias d:
+          shape.defineFunctionTypeAlias(d);
+
+        case TopLevelVariableDeclaration d:
+          shape.defineTopLevelVariable(d);
+      }
+    }
+  }
+
+  // TODO: is the opposite of 'transitive export', 'direct export'?
+  /// Iterate over the direct exports of this library and merge the set of
+  /// exported shapes of those libraries into this one.
+  /// Returns [true] if the shape of [library] has been changed as a result.
```

Comment:
It might be cheaper to propagate around the `export` statements... so updating `library.exports` based on what the libraries being exported have in `exportedLibrary.exports`.

It's a bit more complicated, but this is why the `NamespaceFilter` has the merge/apply logic.

I think we can even be able to track which `LibraryShape` objects have had their `.exports` property modified in the first iteration. Because then we only need to further propogate those in following iterations (or something like that).


----------------

The idea with propogating the `export` declaration is that there are much fewer of them than there are of the `definedShapes`.

I think that if we propogate `export` declarations, we can create `exportedShapes` in a single pass afterwards.

note. maybe a good package to test this on would be something like `package:googleapis`, it's massive and adds a lot of symbols without breaking the API.